[#38833] BCF API Comments Service

Kharonus opened the pull request:
- https://community.openproject.org/work_packages/38833
- added requests tests
- added check for expected body length
- added representer and api for bcf_comments
- added create and set attributes service for bcf_comments
- added migration for comment hierarchy
- added create and update service for journals

In this PR the Comments endpoints of the BCF API v2.1 are added to our stack. In addition, some lightweight changes were made, when the added code interfered with existing api code.

Kharonus committed:
[#38833] BCF API Comments Service

- https://community.openproject.org/work_packages/38833
- added requests tests
- added check for expected body length
- added representer and api for bcf_comments
- added create and set attributes service for bcf_comments
- added migration for comment hierarchy
- added create and update service for journals

Kharonus committed:
[#38833] cosmetic changes

- removed utf-8 encoding flag
- extracted duplicate code in bcf comments api

oliverguenther commented on modules/bim/app/models/bim/bcf/non_existent_comment.rb:
```diff
@@ -0,0 +1,4 @@
+module Bim::Bcf
+  class NonExistentComment < Comment
```
To avoid this from being saved to the DB by accident, add a method

```ruby
def readonly?
  true
end
```

Kharonus replied:
done, good catch

oliverguenther commented on modules/bim/app/models/bim/bcf/non_existent_viewpoint.rb:
```diff
@@ -0,0 +1,4 @@
+module Bim::Bcf
+  class NonExistentViewpoint < Viewpoint
```
Same here

oliverguenther commented on modules/bim/app/contracts/bim/bcf/comments/update_contract.rb:
```diff
@@ -0,0 +1,54 @@
+      private
+
+      def user_allowed_to_update
+        errors.add :base, :error_unauthorized unless @user.allowed_to?(:manage_bcf, model.issue.work_package.project)
```
You can move all these validations into the base contract and avoid duplicating them between the create and update contracts

Kharonus replied:
`BaseContract` is a quite generic class. So, how would you extract the logic? passing specific information as a parameter, like

```
def user_allowed_to?(permission, entity)
  errors.add :base, :error_unauthorized unless @user.allowed_to?(permission, entity)
end
```

Kharonus replied:
extracted to base contract

oliverguenther commented on modules/bim/app/controllers/bim/bcf/api/v2_1/comments/api.rb:
```diff
@@ -0,0 +1,93 @@
+            if viewpoint_uuid == nil
+              nil
+            else
+              @issue.viewpoints.find_by(uuid: viewpoint_uuid) || ::Bim::Bcf::NonExistentViewpoint.new
```
I'm a bit torn with this approach. I can see why you did it: You get validations in the contract and get all errors returned to the user whereas failing the finder would fail the API request with a single error instead.

But you have to instantiate a new record subclass that feels a bit unnecessary. I don't have a better idea, however.

Kharonus replied:
I could fail here but then I would kind of circumvent the role of the contract for this route. This is mostly a design decision. From functionality I (and the API by specification) could live with both.

oliverguenther commented on modules/bim/app/controllers/bim/bcf/api/v2_1/comments/api.rb:
```diff
@@ -0,0 +1,93 @@
+
+        route_param :comment_guid, regexp: /\A[a-f0-9\-]+\z/ do
+          after_validation do
+            @comment = all_comments.find_by!(uuid: params[:comment_guid])
```
Prefer to use `declared_params[:comment_guuid]` as that implies you have declared and validated it. It makes no difference within an `after_validation` block but will prevent errors in case of careless refactorings.

Kharonus replied:
right, thanks for pointing out. changed!

oliverguenther commented on modules/bim/app/services/bim/bcf/comments/update_service.rb:
```diff
@@ -0,0 +1,49 @@
+      def update_journal(journal, comment)
+        ::Journals::UpdateService.new(user: user,
+                                      model: journal,
+                                      contract_class: ::EmptyContract)
```
Why do you need the EmptyContract here?

Kharonus replied:
Updating a journal's notes requires no further validation that could be put into a contract. The alternative is having an empty class `::Journals::UpdateContract` to satisfy the default convention. Giving an empty contract to the update service call make it more clear IMHO, that this does not need further validation.

oliverguenther reviewed (changes requested):
Mostly LGTM 👏  I only have a few remarks.

Kharonus committed:
[#38833] Fixed PR issues

- fixed some rubocop issues
- added parameter declaration to comments api
- added readonly flag to readonly records

Kharonus committed:
[#38833] Extracted comments base contract
